### PR TITLE
Stacked ner

### LINKED ISSUE #300: Error running the trainer with ner_duckling component
**rasa NLU version** (e.g. `0.7.3`):
0.8.0a11
**Used backend / pipeline** (`mitie`, `spacy_sklearn`, ...):
"nlp_mitie","ner_duckling"

**Operating system** (windows, osx, ...):

**Issue**:
Traceback (most recent call last):
  File "/usr/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/areyes/rasa_nlu/rasa_nlu/train.py", line 81, in <module>
    do_train(config)
  File "/home/areyes/rasa_nlu/rasa_nlu/train.py", line 70, in do_train
    trainer.validate()
  File "rasa_nlu/model.py", line 164, in validate
    raise Exception("Failed to validate at component '{}'. {}".format(component.name, e.message))
Exception: Failed to validate at component 'ner_duckling'. Couldn't fill argument 'entities' :(

**Content of configuration file** (if used & relevant):
```json
{
  "pipeline":["nlp_mitie","ner_duckling"],
  "mitie_file": "./data/total_word_feature_extractor.dat",
  "path" : "./models",
  "num_threads":4,
  "logs":"logs",
  "port":5001,
  "duckling_processing_mode": "append",
  "data" : "./data/charterbot/charterBotTrn.json",
  "token":"3cXuFQdNBae2l/iInieT7Ud3mDVXwZUY"
}
```

### PR DESCRIPTION
**Pull request by twerkmeister**
**Proposed changes**:
- fixes #300, #310 

@tmbo do you think at this point this implementation suffices or should we go directly for a more general solution, e.g. a pipeline class that defines certain requirements for the context?

**Status**:
- [x] ready for code review
- [x] there are tests for the functionality
- [x] documentation updated
- [x] changelog updated

### CODE REVIEW

**Comment by tmbo** [comment]
I think we can use the default output attributes from the Interpreter for that (so initialize the context to these instead of being empty. we only need to be carful with the text attribute as this is set to `""` in the default output but should be set to the incomming message in the context.)

I think this solves #300 partly as it still leaves us with the ordering issue. I think it would be good to move the merging to the abstract NER class at least for the cases where we can reuse it (e.g. append without creating duplicates).

**Comment by twerkmeister** [comment]
Ok, thanks for the suggestion!

We could now use the duckling component as a first NER component. What part of the ordering issue are you still seeing as remaining?

**Comment by tmbo** [comment]
If you put another NER component after the duckling component, e.g. `ner_crf` it will override the duckling entities instead of combining the results.

**Comment by amn41** [comment]
when the final behaviour is decided please also update #325 docs

**Comment by twerkmeister** [comment]
Updated according to the discussion in https://github.com/golastmile/rasa_nlu/issues/310. Please share some feedback on the implementation @tmbo :)

File: rasa_nlu/config.py [hunk 1]
```diff
@@ -32,7 +32,7 @@
     "max_number_of_ngrams": 7,
     "pipeline": [],
     "response_log": "logs",
-    "duckling_processing_mode": "append",
+    "duckling_dimensions": [],
```
Comment (tmbo) [inline rasa_nlu/config.py]: should the default be "no dimension"?

File: rasa_nlu/extractors/duckling_extractor.py [hunk 2]
```diff
@@ -10,18 +10,19 @@
...
 from rasa_nlu.model import Metadata
 
 
 if typing.TYPE_CHECKING:
     from duckling import DucklingWrapper
 
 
-DUCKLING_PROCESSING_MODES = ["replace", "append"]
+DUCKLING_DIMENSIONS = ["time", "temperature", "number", "ordinal", "distance", "volume",
+                       "amount-of-money", "duration", "email", "url", "phone-number"]
```
Comment (tmbo) [inline rasa_nlu/extractors/duckling_extractor.py]: There is a list of available dimensions here, https://github.com/FraBle/python-duckling/blob/master/duckling/dim.py. I think we should use these constants. (and i think you missed timezone)

File: rasa_nlu/extractors/duckling_extractor.py [hunk 3]
```diff
@@ -32,10 +33,13 @@ class DucklingExtractor(Component):
...
 
-    def __init__(self, duckling_processing_mode, duckling=None):
+    def __init__(self, duckling_dimensions=DUCKLING_DIMENSIONS, duckling=None):
         # type: (Text, Optional[DucklingWrapper]) -> None
 
-        self.duckling_processing_mode = duckling_processing_mode
+        # If duckling dimensions == [] or == None
+        if not duckling_dimensions:
+            duckling_dimensions = DUCKLING_DIMENSIONS
+        self.duckling_dimensions = duckling_dimensions
```
Comment (tmbo) [inline rasa_nlu/extractors/duckling_extractor.py]: ternary op?

File: rasa_nlu/extractors/duckling_extractor.py [hunk 4]
```diff
@@ -44,12 +48,13 @@ def required_packages(cls):
         return ["duckling"]
 
     @classmethod
-    def create(cls, duckling_processing_mode):
-        if duckling_processing_mode not in DUCKLING_PROCESSING_MODES:
-            raise ValueError("Invalid duckling processing mode. Got '{}'. Allowed: {}".format(
-                duckling_processing_mode, ", ".join(DUCKLING_PROCESSING_MODES)))
+    def create(cls, duckling_dimensions=DUCKLING_DIMENSIONS):
```
Comment (tmbo) [inline rasa_nlu/extractors/duckling_extractor.py]: shouldn't take default arguments as these are filled using reflection and the default will always be overwritten

File: rasa_nlu/extractors/duckling_extractor.py [hunk 5]
```diff
@@ -68,34 +73,27 @@ def pipeline_init(self, language):
...
+
+        extracted = self.add_extractor_name(extracted)
         return {
-            "entities": entities
+            "entities": entities.extend(extracted)
         }
 
     @classmethod
-    def load(cls, duckling_processing_mode):
+    def load(cls, duckling_dimensions=DUCKLING_DIMENSIONS):
```
Comment (tmbo) [inline rasa_nlu/extractors/duckling_extractor.py]: `duckling_dimensions` should be persisted to the metadata in the persist call. Shouldn't take default arg here.

File: rasa_nlu/model.py [hunk 6]
```diff
@@ -226,7 +227,7 @@ def __init__(self, pipeline, context, config, meta=None):
         # type: (List[Component], Dict[Text, Any], Dict[Text, Any], Optional[Metadata]) -> None
 
         self.pipeline = pipeline
-        self.context = context
+        self.context = self.default_output_attributes.update(context)
```
Comment (tmbo) [inline rasa_nlu/model.py]: Does this work? wont the call a) update the default output attributes and b) return nothing?

File: rasa_nlu/config.py (hunk 1 above)
Comment (twerkmeister) [inline rasa_nlu/config.py]: It's actually all modules. I could also import the list from the duckling module to make it more explicit, but then if the duckling module ever imports `config` then we have a loop :curly_loop:

Comment (tmbo) [inline rasa_nlu/config.py]: Alright, gotcha. But then I'd go for `None`

**Pull request merged**